File: basic_dsa_thirty.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Check if a Number is Ugly Number
# An Ugly number is a number whose only prime factors are 2, 3, or 5
# Example: 6 = 2 × 3 (ugly), 8 = 2 × 2 × 2 (ugly), 14 = 2 × 7 (not ugly, has factor 7)

def is_ugly_number(n):
    """
    Check if a number is Ugly number
    An Ugly number has only prime factors 2, 3, or 5
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    if n <= 0:
        return False
    if n == 1:
        return True  # 1 is considered ugly
    
    # Remove all factors of 2, 3, and 5
    while n % 2 == 0:
        n //= 2
    while n % 3 == 0:
        n //= 3
    while n % 5 == 0:
        n //= 5
    
    # If after removing 2, 3, 5 factors, we get 1, it's ugly
    return n == 1


def get_ugly_factors(n):
    """
    Get prime factors of a number (only 2, 3, 5 if ugly)
    Returns list of factors
    Time Complexity: O(log n)
    Space Complexity: O(log n)
    """
    if n <= 0:
        return []
    
    factors = []
    num = n
    
    # Check factors of 2, 3, 5
    for prime in [2, 3, 5]:
        while num % prime == 0:
            factors.append(prime)
            num //= prime
    
    return factors if num == 1 else None  # None if has other factors


def get_ugly_numbers_in_range(start, end):
    """
    Find all Ugly numbers in a given range
    Time Complexity: O((end - start) * log n)
    Space Complexity: O(k) where k is the number of Ugly numbers
    """
    ugly_numbers = []
    
    for num in range(start, end + 1):
        if is_ugly_number(num):
            ugly_numbers.append(num)
    
    return ugly_numbers


# Problem 2: Find Sum of First N Odd Numbers
# Sum of first n odd numbers = n²
# Example: 1 + 3 + 5 + 7 = 16 = 4²

def sum_of_odd_numbers_iterative(n):
    """
    Find sum of first n odd numbers using iteration
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if n <= 0:
        return 0
    
    total = 0
    odd = 1
    
    for i in range(n):
        total += odd
        odd += 2
    
    return total


def sum_of_odd_numbers_formula(n):
    """
    Find sum of first n odd numbers using formula
    Sum = n²
    Time Complexity: O(1)
    Space Complexity: O(1)
    """
    if n <= 0:
        return 0
    return n * n


def sum_of_odd_numbers_recursive(n):
    """
    Find sum of first n odd numbers using recursion
    Time Complexity: O(n)
    Space Complexity: O(n) due to recursion stack
    """
    if n <= 0:
        return 0
    if n == 1:
        return 1
    
    # nth odd number = 2n - 1
    nth_odd = 2 * n - 1
    return nth_odd + sum_of_odd_numbers_recursive(n - 1)


def get_first_n_odd_numbers(n):
    """
    Get list of first n odd numbers
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    if n <= 0:
        return []
    
    odd_numbers = []
    odd = 1
    
    for i in range(n):
        odd_numbers.append(odd)
        odd += 2
    
    return odd_numbers


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Check if a Number is Ugly Number")
    print("=" * 60)
    
    # Test Case 1 - Ugly number 6
    num1 = 6
    result1 = is_ugly_number(num1)
    factors1 = get_ugly_factors(num1)
    print(f"Number: {num1}")
    if factors1:
        print(f"Prime factors: {factors1}")
    print(f"Is Ugly number: {result1}")
    
    # Test Case 2 - Ugly number 8
    num2 = 8
    result2 = is_ugly_number(num2)
    factors2 = get_ugly_factors(num2)
    print(f"\nNumber: {num2}")
    if factors2:
        print(f"Prime factors: {factors2}")
    print(f"Is Ugly number: {result2}")
    
    # Test Case 3 - Not Ugly number 14
    num3 = 14
    result3 = is_ugly_number(num3)
    factors3 = get_ugly_factors(num3)
    print(f"\nNumber: {num3}")
    if factors3:
        print(f"Prime factors: {factors3}")
    else:
        print("Has prime factors other than 2, 3, 5")
    print(f"Is Ugly number: {result3}")
    
    # Test Case 4 - Ugly number 1
    num4 = 1
    result4 = is_ugly_number(num4)
    print(f"\nNumber: {num4}")
    print(f"Is Ugly number: {result4}")
    
    # Test Case 5 - Find Ugly numbers in range
    start = 1
    end = 30
    ugly_nums = get_ugly_numbers_in_range(start, end)
    print(f"\nUgly numbers in range [{start}, {end}]:")
    print(ugly_nums)
    print(f"Count: {len(ugly_nums)}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Find Sum of First N Odd Numbers")
    print("=" * 60)
    
    # Test Case 1 - Sum of first 4 odd numbers
    n1 = 4
    result1_iter = sum_of_odd_numbers_iterative(n1)
    result1_formula = sum_of_odd_numbers_formula(n1)
    result1_rec = sum_of_odd_numbers_recursive(n1)
    odd_nums1 = get_first_n_odd_numbers(n1)
    print(f"n = {n1}")
    print(f"First {n1} odd numbers: {odd_nums1}")
    print(f"Sum (iterative): {result1_iter}")
    print(f"Sum (formula): {result1_formula}")
    print(f"Sum (recursive): {result1_rec}")
    print(f"Verification: {n1}² = {n1 * n1}")
    
    # Test Case 2 - Sum of first 10 odd numbers
    n2 = 10
    result2 = sum_of_odd_numbers_formula(n2)
    odd_nums2 = get_first_n_odd_numbers(n2)
    print(f"\nn = {n2}")
    print(f"First {n2} odd numbers: {odd_nums2}")
    print(f"Sum: {result2}")
    print(f"Verification: {n2}² = {n2 * n2}")
    
    # Test Case 3 - Sum of first 1 odd number
    n3 = 1
    result3 = sum_of_odd_numbers_formula(n3)
    print(f"\nn = {n3}")
    print(f"Sum: {result3}")
    
    # Test Case 4 - Sum of first 5 odd numbers
    n4 = 5
    result4 = sum_of_odd_numbers_formula(n4)
    odd_nums4 = get_first_n_odd_numbers(n4)
    print(f"\nn = {n4}")
    print(f"First {n4} odd numbers: {odd_nums4}")
    print(f"Sum: {result4}")
    
    # Test Case 5 - Large n
    n5 = 20
    result5 = sum_of_odd_numbers_formula(n5)
    print(f"\nn = {n5}")
    print(f"Sum of first {n5} odd numbers: {result5}")
    print(f"Verification: {n5}² = {n5 * n5}")

File: basic_dsa_thirty.java
/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */
import java.util.*;

public class basic_dsa_thirty {
    
    // Problem 1: Check if a Number is Ugly Number
    // An Ugly number is a number whose only prime factors are 2, 3, or 5
    // Example: 6 = 2 × 3 (ugly), 8 = 2 × 2 × 2 (ugly), 14 = 2 × 7 (not ugly, has factor 7)
    
    /**
     * Check if a number is Ugly number
     * An Ugly number has only prime factors 2, 3, or 5
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     */
    public static boolean isUglyNumber(int n) {
        if (n <= 0) {
            return false;
        }
        if (n == 1) {
            return true; // 1 is considered ugly
        }
        
        // Remove all factors of 2, 3, and 5
        while (n % 2 == 0) {
            n /= 2;
        }
        while (n % 3 == 0) {
            n /= 3;
        }
        while (n % 5 == 0) {
            n /= 5;
        }
        
        // If after removing 2, 3, 5 factors, we get 1, it's ugly
        return n == 1;
    }
    
    /**
     * Get prime factors of a number (only 2, 3, 5 if ugly)
     * Returns list of factors or null if has other factors
     * Time Complexity: O(log n)
     * Space Complexity: O(log n)
     */
    public static ArrayList<Integer> getUglyFactors(int n) {
        if (n <= 0) {
            return null;
        }
        
        ArrayList<Integer> factors = new ArrayList<>();
        int num = n;
        
        // Check factors of 2, 3, 5
        int[] primes = {2, 3, 5};
        for (int prime : primes) {
            while (num % prime == 0) {
                factors.add(prime);
                num /= prime;
            }
        }
        
        return num == 1 ? factors : null;
    }
    
    /**
     * Find all Ugly numbers in a given range
     * Time Complexity: O((end - start) * log n)
     * Space Complexity: O(k) where k is the number of Ugly numbers
     */
    public static ArrayList<Integer> getUglyNumbersInRange(int start, int end) {
        ArrayList<Integer> uglyNumbers = new ArrayList<>();
        
        for (int num = start; num <= end; num++) {
            if (isUglyNumber(num)) {
                uglyNumbers.add(num);
            }
        }
        
        return uglyNumbers;
    }
    
    // Problem 2: Find Sum of First N Odd Numbers
    // Sum of first n odd numbers = n²
    // Example: 1 + 3 + 5 + 7 = 16 = 4²
    
    /**
     * Find sum of first n odd numbers using iteration
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static long sumOfOddNumbersIterative(int n) {
        if (n <= 0) {
            return 0;
        }
        
        long total = 0;
        int odd = 1;
        
        for (int i = 0; i < n; i++) {
            total += odd;
            odd += 2;
        }
        
        return total;
    }
    
    /**
     * Find sum of first n odd numbers using formula
     * Sum = n²
     * Time Complexity: O(1)
     * Space Complexity: O(1)
     */
    public static long sumOfOddNumbersFormula(int n) {
        if (n <= 0) {
            return 0;
        }
        return (long) n * n;
    }
    
    /**
     * Find sum of first n odd numbers using recursion
     * Time Complexity: O(n)
     * Space Complexity: O(n) due to recursion stack
     */
    public static long sumOfOddNumbersRecursive(int n) {
        if (n <= 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        
        // nth odd number = 2n - 1
        int nthOdd = 2 * n - 1;
        return nthOdd + sumOfOddNumbersRecursive(n - 1);
    }
    
    /**
     * Get list of first n odd numbers
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     */
    public static ArrayList<Integer> getFirstNOddNumbers(int n) {
        ArrayList<Integer> oddNumbers = new ArrayList<>();
        if (n <= 0) {
            return oddNumbers;
        }
        
        int odd = 1;
        for (int i = 0; i < n; i++) {
            oddNumbers.add(odd);
            odd += 2;
        }
        
        return oddNumbers;
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Check if a Number is Ugly Number");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Ugly number 6
        int num1 = 6;
        boolean result1 = isUglyNumber(num1);
        ArrayList<Integer> factors1 = getUglyFactors(num1);
        System.out.println("Number: " + num1);
        if (factors1 != null) {
            System.out.println("Prime factors: " + factors1);
        }
        System.out.println("Is Ugly number: " + result1);
        
        // Test Case 2 - Ugly number 8
        int num2 = 8;
        boolean result2 = isUglyNumber(num2);
        ArrayList<Integer> factors2 = getUglyFactors(num2);
        System.out.println("\nNumber: " + num2);
        if (factors2 != null) {
            System.out.println("Prime factors: " + factors2);
        }
        System.out.println("Is Ugly number: " + result2);
        
        // Test Case 3 - Not Ugly number 14
        int num3 = 14;
        boolean result3 = isUglyNumber(num3);
        ArrayList<Integer> factors3 = getUglyFactors(num3);
        System.out.println("\nNumber: " + num3);
        if (factors3 != null) {
            System.out.println("Prime factors: " + factors3);
        } else {
            System.out.println("Has prime factors other than 2, 3, 5");
        }
        System.out.println("Is Ugly number: " + result3);
        
        // Test Case 4 - Ugly number 1
        int num4 = 1;
        boolean result4 = isUglyNumber(num4);
        System.out.println("\nNumber: " + num4);
        System.out.println("Is Ugly number: " + result4);
        
        // Test Case 5 - Find Ugly numbers in range
        int start = 1;
        int end = 30;
        ArrayList<Integer> uglyNums = getUglyNumbersInRange(start, end);
        System.out.println("\nUgly numbers in range [" + start + ", " + end + "]:");
        System.out.println(uglyNums);
        System.out.println("Count: " + uglyNums.size());
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Problem 2: Find Sum of First N Odd Numbers");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Sum of first 4 odd numbers
        int n1 = 4;
        long result1_sum = sumOfOddNumbersIterative(n1);
        long result1_formula = sumOfOddNumbersFormula(n1);
        long result1_rec = sumOfOddNumbersRecursive(n1);
        ArrayList<Integer> oddNums1 = getFirstNOddNumbers(n1);
        System.out.println("n = " + n1);
        System.out.println("First " + n1 + " odd numbers: " + oddNums1);
        System.out.println("Sum (iterative): " + result1_sum);
        System.out.println("Sum (formula): " + result1_formula);
        System.out.println("Sum (recursive): " + result1_rec);
        System.out.println("Verification: " + n1 + "² = " + (n1 * n1));
        
        // Test Case 2 - Sum of first 10 odd numbers
        int n2 = 10;
        long result2_sum = sumOfOddNumbersFormula(n2);
        ArrayList<Integer> oddNums2 = getFirstNOddNumbers(n2);
        System.out.println("\nn = " + n2);
        System.out.println("First " + n2 + " odd numbers: " + oddNums2);
        System.out.println("Sum: " + result2_sum);
        System.out.println("Verification: " + n2 + "² = " + (n2 * n2));
        
        // Test Case 3 - Sum of first 1 odd number
        int n3 = 1;
        long result3_sum = sumOfOddNumbersFormula(n3);
        System.out.println("\nn = " + n3);
        System.out.println("Sum: " + result3_sum);
        
        // Test Case 4 - Sum of first 5 odd numbers
        int n4 = 5;
        long result4_sum = sumOfOddNumbersFormula(n4);
        ArrayList<Integer> oddNums4 = getFirstNOddNumbers(n4);
        System.out.println("\nn = " + n4);
        System.out.println("First " + n4 + " odd numbers: " + oddNums4);
        System.out.println("Sum: " + result4_sum);
        
        // Test Case 5 - Large n
        int n5 = 20;
        long result5_sum = sumOfOddNumbersFormula(n5);
        System.out.println("\nn = " + n5);
        System.out.println("Sum of first " + n5 + " odd numbers: " + result5_sum);
        System.out.println("Verification: " + n5 + "² = " + ((long) n5 * n5));
    }
}
